File: eratosthenes/eratosthenes2.cc
#include "eratosthenes2.h"

#include <cmath>

namespace {
#include "constants.h"
}  // namespace

void Eratosthenes2::generate(const int64 x) {
  flags_.clear();
  if (x > 10000000000)
    return;

  int64 size = x / 30 + (x % 30 != 0);
  flags_.resize(size, 0xff);
  if (int r = x % 30) {
    if (r <= 1)       flags_[size - 1] = 0x0;
    else if (r <= 7)  flags_[size - 1] = 0x1;
    else if (r <= 11) flags_[size - 1] = 0x3;
    else if (r <= 13) flags_[size - 1] = 0x7;
    else if (r <= 17) flags_[size - 1] = 0xf;
    else if (r <= 19) flags_[size - 1] = 0x1f;
    else if (r <= 23) flags_[size - 1] = 0x3f;
    else if (r <= 29) flags_[size - 1] = 0x7f;
  }

  flags_[0] = 0xfe;
  const uint64 sqrt_x = std::ceil(std::sqrt(x) + 0.1);
  const uint64 sqrt_xi = sqrt_x / 30 + 1;
  for (uint64 i = 0; i < sqrt_xi; ++i) {
    for (uint8 flags = flags_[i]; flags; flags &= flags - 1) {
      uint8 lsb = flags & (-flags);
      int ibit = BitToIndex(lsb);
      const int32 m = kMod30[ibit];
      int32 pm = 30 * i + 2 * m;
      for (uint64 j = i * pm + (m * m) / 30, k = ibit; j < flags_.size();
           j += i * C1[k] + C0[ibit][k], k = (k + 1) & 7) {
        flags_[j] &= kMask[ibit][k];
      }
    }
  }
}

int64 Eratosthenes2::count() {
  if (flags_.empty())
    return -1;

  int64 ret = 3;  // count 2, 3, 5
  for (uint8 f : flags_) {
    ret += PopCnt(f);
  }
  return ret;
}

File: eratosthenes/rust/src/eratosthenes/eratosthenes2.rs
use crate::bits::Bytes;

use super::{Eratosthenes, BIT_MASKS, C0, C1, MOD30};

pub struct Eratosthenes2 {
    flags: Bytes,
}

impl Eratosthenes2 {
    pub fn new() -> Eratosthenes2 {
        Eratosthenes2 {
            flags: Bytes::new(0, 0),
        }
    }
}

impl Eratosthenes for Eratosthenes2 {
    fn version(&self) -> i32 {
        2
    }

    fn generate(&mut self, x: usize) {
        let r = x % 30;
        let size = x / 30 + if r == 0 { 0 } else { 1 };
        let mut flags = Bytes::new(size, 0xff);

        flags.set_byte(
            size - 1,
            match r {
                0 => 0xff,
                1 => 0x00,
                2..=7 => 0x01,
                8..=11 => 0x03,
                12..=13 => 0x07,
                14..=17 => 0x0f,
                18..=19 => 0x1f,
                20..=23 => 0x3f,
                _ => 0x7f,
            },
        );

        flags.unset_bit(0, 0); // 1 is not a prime
        let sqr_x = (x as f64).sqrt().ceil() as usize;
        let sqr_xi = sqr_x / 30 + 1;
        for i in 0..sqr_xi {
            let mut bits = flags.get_byte(i);
            while bits != 0 {
                let lsb = bits & (!bits + 1);
                let ibit = (lsb - 1).count_ones() as usize;
                let m = MOD30[ibit];
                let pm = 30 * i + 2 * m;

                let mut j = i * pm + (m * m) / 30;
                let mut k = ibit;
                while j < size {
                    flags.mask(j, BIT_MASKS[ibit][k]);
                    j += i * C1[k] + C0[ibit][k];
                    k = (k + 1) & 7;
                }
                bits &= bits - 1;
            }
        }
        self.flags = flags;
    }

    fn generate_range(&mut self, _x: usize, _y: usize) {
        eprintln!("Does not support ranged generation.");
    }

    fn count(&self) -> usize {
        // 2, 3, and 5 are counted separately.
        3 + self.flags.bit_count()
    }
}
